res/controllers/M-Audio-Xponent-Advanced-scripts.js

```diff
@@ -431,60 +473,57 @@ MaudioXponent.effectButton = function(channel, control, value, status, group) {
+            if (focusedEffect === button) {
+                // Toggle focused Effect
+                group = "[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "_Effect" + focusedEffect + "]";
+                var enabled = !engine.getValue(group, "enabled");
+                engine.setValue(group, "enabled", enabled);
+            } else {
+                // Focus effect
+                engine.setValue(group, "focused_effect", button);
+                engine.setValue(group, "show_focus", true);
+            }
+        }
     }
 };
 
-MaudioXponent.effectParameter = function(channel, control, value, status, group) {
-    // script.midiDebug(channel, control, value, status, group);
+MaudioXponent.effectKnob = function(channel, control, value) {
     var knob = control - 0x0B;
     var scaledValue = value / 0x7F;
 
-    if (MaudioXponent.state.focusedEffectUnit != 0) {
+    if (MaudioXponent.state.focusedEffectUnit === 0) {
         if (MaudioXponent.rightDeck.shift) {
-            // Manipulate focused EffectUnit parameters
-            if (knob == 4) {
-                // Wet / Dry
-                engine.setParameter("[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "]", "mix", scaledValue);
-            } else {
-                // Other parameter
-                engine.setParameter("[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "_Effect1]", "parameter" + knob, scaledValue);
-            }
+            // Manipulate Wet/Dry mix
+            engine.setParameter("[EffectRack1_EffectUnit" + knob + "]", "mix", scaledValue);
         } else {
-            // Manipulate focused EffectUnit meta/super knobs
-            if (knob == 4) {
-                // Super-knob
-                group = "[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "]";
-                engine.setParameter(group, "super1", scaledValue);
-            } else {
-                // Meta-knob
-                group = "[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "_Effect" + knob + "]";
-                engine.setParameter(group, "meta", scaledValue);
-            }
+            // Manipulate Unit Super-knobs
+            engine.setParameter("[EffectRack1_EffectUnit" + knob + "]", "super1", scaledValue);
         }
     } else {
-        if (MaudioXponent.rightDeck.shift) {
-            // Manipulate Wet/Dry mix
-            group = "[EffectRack1_EffectUnit" + knob + "]";
-            engine.setParameter(group, "mix", scaledValue);
+        // An effect has focus
+        var focusedEffect = engine.getValue("[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "]", "focused_effect");
+        
+        if (focusedEffect === 0) {
+            // Manipulate focused EffectUnit meta knobs
+            engine.setParameter("[EffectRack1_EffectUnit" + MaudioXponent.state.focusedEffectUnit + "_Effect" + knob + "]", "meta", scaledValue);
```

Comment:
I'm assuming at least some skin out there could be built to allow an arbitrary number of effects, so I'm making no assumptions about the maximum here. If someone makes a skin with four effects, this will work with it, otherwise you have no way of actually loading an effect into the invisible fourth slot anyway, so you can't really hurt yourself with this.